crates/red_knot/src/program/mod.rs

```diff
@@ -52,14 +52,18 @@ impl Program {
         &mut self.workspace
     }
 
-    #[allow(clippy::unnecessary_wraps)]
     fn with_db<F, T>(&self, f: F) -> Result<T, Cancelled>
     where
-        F: FnOnce(&Program) -> T + UnwindSafe,
+        F: FnOnce(&Program) -> T + std::panic::UnwindSafe,
     {
-        // TODO: Catch in `Caancelled::catch`
-        //  See https://salsa.zulipchat.com/#narrow/stream/145099-general/topic/How.20to.20use.20.60Cancelled.3A.3Acatch.60
-        Ok(f(self))
+        // (UN?)SAFETY: I don't think this is a 100% safe, lol. But it's what everyone else does.
+        // The salsa storage is guaranteed to be unwind safe (or at least, has been designed to be unwind
+        // safe for Salsa-exceptions). But this is not strictly guaranteed for any user code in `Program`.
+        // For example, multi-threaded inside `Program` that calls a query could panic and unwind, before
+        // it can complete and transitions into a valid state.
+        // To me, this seems like a design flaw in salsa, see https://salsa.zulipchat.com/#narrow/stream/145099-general/topic/How.20to.20use.20.60Cancelled.3A.3Acatch.60
+        let db = &AssertUnwindSafe(self);
```

Comment:
I don't have a ton of context here, but I'm not too concerned about this personally. Unwind Safety is not the same as Safety in Rust-speak. Like, there's no UB risk here. Violating unwind safety can lead to logic bugs around mutation or inconsistent state, but is never supported to lead to UB. There have been rumblings about [deprecating unwind safety altogether](https://internals.rust-lang.org/t/pre-rfc-deprecating-unwindsafe/15974).

Reply:
With that said, is this using panicking as a means of error handling? That seems like bad juju, but like, I have no context at all on which to judge that design.

Reply:
Thanks @BurntSushi. That's very helpful. I share some of the sentiment from the linked thread. It was very unclear to me how big of a problem it is. The compiler erroring makes it extremely scary. I now feel less bad about this.

> With that said, is this using panicking as a means of error handling? 

Unfortunately yes. Salsa uses panics to abort queries when some input has changed to not waste time computing outdated results to error out and to recover from cyclic queries. I would rather prefer a `Result`, but I can see how this would hurt ergonomics quite a bit. But maybe something to bring up again, although I think it's an unlikely change to be accepted because rust-analyzer would have to rewrite all their code to Result's.